ext/standard/basic_functions.c

```diff
@@ -1214,18 +1216,38 @@ PHP_FUNCTION(flush)
 /* {{{ Delay for a given number of seconds */
 PHP_FUNCTION(sleep)
 {
-	zend_long num;
+	double seconds;
 
 	ZEND_PARSE_PARAMETERS_START(1, 1)
-		Z_PARAM_LONG(num)
+		Z_PARAM_DOUBLE(seconds)
 	ZEND_PARSE_PARAMETERS_END();
 
-	if (num < 0) {
+	if (isnan(seconds) || seconds < 0) {
 		zend_argument_value_error(1, "must be greater than or equal to 0");
 		RETURN_THROWS();
 	}
 
-	RETURN_LONG(php_sleep((unsigned int)num));
+#if HAVE_NANOSLEEP
+	struct timespec php_req, php_rem;
+
+	if (seconds < LONG_MAX) { // prevent time_t overflow, equals to about 68 years
+		php_req.tv_sec = (time_t) floor(seconds);
+		php_req.tv_nsec = (long) (fmod(seconds, 1.0) * NANOS_IN_SEC);
+	} else {
+		php_req.tv_sec = (time_t) LONG_MAX;
+		php_req.tv_nsec = 0;
+	}
+
+	if (!nanosleep(&php_req, &php_rem)) {
+		RETURN_DOUBLE(0);
+	} else if (errno == EINTR) {
+		RETURN_DOUBLE((double)php_rem.tv_sec + (double)php_rem.tv_nsec / NANOS_IN_SEC);
+	}
+#else
```

Comment:
```suggestion
#elif HAVE_USLEEP
#if HAVE_USLEEP
	usleep(seconds * 1000000 <= UINT_MAX ? (unsigned int) (seconds * 1000000) : UINT_MAX));
#else
```

Something along those lines - we can use usleep if it exists so that sleep(0.1) would work reasonably well.

Looking at the usleep implementation, though, it doesn't check if the value is too large, either.

Reply:
`usleep` currently does not return "the number of seconds left to sleep, if the call was interrupted by a signal handler"

`usleep` does return however if interrupted or not, thus we can detect an interrupt and very easily calculate the remaining time to sleep

I belive the current return value definition is also based on the standard/underlaying C `sleep` without further consideration. As the return is very rarely used (see PR description why, at least on Windows it returned one fixed value) what about changing/abstracting this function further to return:
- `false` if sleep was not interrupted (previously `0`, thus non strict return comparison with still work)
- `true` if interrupted during sleep

if user need the remaining time, he/she can very easily measure it with much more control

@nikic @cmb69 @derickr @KalleZ @TysonAndre @moliata wdyt?

Reply:
It'd be a backwards incompatible change and wouldn't make sense out of a major version release.

Also, the behavior is already documented and I don't feel like changing that is worth the confusion and complications it adds to the code that *did* use the return for supporting php 8 and 9

Reply:
Thanks for your feedback, I have now added usleep fallback.